src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -439,70 +442,80 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
         /// <returns>Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.</returns>
         internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasProperties)
         {
-            if (CommonUtil.IsUnsupportedType(type))
-            {
-                throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
-            }
+            ODataTypeInfo typeInfo = GetODataTypeInfo(type);
 
-            string typeName = type.ToString();
-            IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
-            bool isEntity = customAttributes.OfType<EntityTypeAttribute>().Any();
-            KeyAttribute dataServiceKeyAttribute = customAttributes.OfType<KeyAttribute>().FirstOrDefault();
-            List<PropertyInfo> keyProperties = new List<PropertyInfo>();
-            PropertyInfo[] properties = ClientTypeUtil.GetPropertiesOnType(type, false /*declaredOnly*/).ToArray();
- 
-            hasProperties = properties.Length > 0;
-            KeyKind currentKeyKind = KeyKind.NotKey;
-            KeyKind newKeyKind = KeyKind.NotKey;
-            foreach (PropertyInfo propertyInfo in properties)
+            if (typeInfo.KeyProperties == null)

```

Comment:
if (typeInfo != null)
{
    return ....;
}

Keep the existing code unchanged.

Reply:
typeinfo will be not null